login/login.c

```diff
@@ -249,24 +219,109 @@ int request_url(const uint8_t service_key[GLOME_MAX_PUBLIC_KEY_LENGTH],
+  UNUSED(config);
+  UNUSED(pamh);
+  va_list args;
+  va_start(args, format);
+  vfprintf(stderr, format, args);
+  va_end(args);
+  fflush(NULL);
+}
+
+void login_syslog(glome_login_config_t* config, pam_handle_t* pamh,
+                  int priority, const char* format, ...) {
+  UNUSED(pamh);
   if (config->options & SYSLOG) {
-    openlog("glome-login", LOG_PID | LOG_CONS, LOG_AUTH);
+    va_list args;
+    va_start(args, format);
+    vsyslog(priority, format, args);
+    va_end(args);
+  }
+}
+
+// read_stdin reads printable characters from stdin into buf. It returns:
+// -1, if it encounters an error while reading
+// -2, if it encounters EOF
+// (buflen-1) if it read buflen-1 characters
+// <(buflen-1), if a newline was read before the buffer was full
+// If the return value is >=0, the buf is NULL-terminated.
+static int read_stdin(char* buf, int buflen) {
+  int i = 0;
+
+  while (i < buflen - 1) {
+    int n = read(STDIN_FILENO, buf + i, 1);
+    if (n < 0) {  // error while reading from stdin
+      perror("ERROR when reading from stdin");
+      return -1;
+    }
+    if (n == 0) {  // EOF
+      return -2;
+    }
+    if (buf[i] == '\n' || buf[i] == '\r') {  // newline
+      break;
+    } else if (buf[i] >= 0x20 && buf[i] <= 0x7e) {
+      // Advance the buffer pointer only if we got a printable character.
+      i++;
+    }
+  }
+  buf[i] = '\0';
+  return i;  // number of characters in the buffer without the NUL byte
+}
+
+static void print_hex(const uint8_t* buf, size_t len) {
+  for (size_t i = 0; i < len; i++) {
+    errorf("%02x", buf[i]);
+  }
+  errorf("\n");
+}
+
+int login_prompt(glome_login_config_t* config, pam_handle_t* pamh,
+                 const char** error_tag, const char* message, char* input,
+                 size_t input_size) {
+  UNUSED(pamh);
+  UNUSED(error_tag);
+
+  puts(message);
+  fputs(PROMPT, stdout);
+  fflush(NULL);
+
+  if (config->input_timeout_sec) {
+    struct sigaction action = {.sa_handler = &timeout_handler};
+    if (sigaction(SIGALRM, &action, NULL) < 0) {
+      perror("error while setting up the handler");
+      // Continue nonetheless as the handler is not critical.
+    }
+    // Set an alarm to prevent waiting for the code indefinitely.
+    alarm(config->input_timeout_sec);
   }
 
+  int bytes_read = read_stdin(input, input_size);
+
+  // Cancel any pending alarms.
+  alarm(0);
+
+  if (bytes_read < 0) {
+    return EXITCODE_IO_ERROR;
+  }
+  if (config->options & VERBOSE) {
+    errorf("debug: stdin: ");
+    print_hex((uint8_t*)input, bytes_read);
+  }
+  return 0;
+}
+#endif
+
+int login_authenticate(glome_login_config_t* config, pam_handle_t* pamh,
+                       const char* prompt_format, const char** error_tag) {
   if (is_zeroed(config->service_key, sizeof config->service_key)) {
     return failure(EXITCODE_PANIC, error_tag, "no-service-key");
   }
+  if (config->url_prefix == NULL) {
+    return failure(EXITCODE_PANIC, error_tag, "no-url-prefix");
```

Comment:
This fixes a bug: before the change, the URL would have been just `(null)/v1/....`. I'd personally have opted for treating an absent prefix as empty, but this works for me, too. Just wanted to make sure that this API change is deliberate.

Reply:
Added empty prefix.

Reply:
Yes, sorry, I missed that part somehow.